src/renderer/Dataframe.js

```diff
@@ -345,69 +441,93 @@ export default class Dataframe {
-            gl.deleteBuffer(this.featureIDBuffer);
-        }
-        const freeObserver = this.freeObserver;
-        Object.keys(this).map(key => {
-            this[key] = null;
-        });
-        this.freed = true;
-        if (freeObserver) {
-            freeObserver(this);
-        }
+    _computePointWidthScale (feature, viz) {
+        const SATURATION_PX = 126;
+        const diameter = Math.min(viz.width.eval(feature) + viz.strokeWidth.eval(feature), SATURATION_PX);
+
+        return diameter / 2 * this.widthScale;
+    }
+
+    _computeLineWidthScale (feature, viz) {
+        const SATURATION_PX = 336;
+        const diameter = Math.min(viz.width.eval(feature), SATURATION_PX);
+
+        return diameter / 2 * this.widthScale;
     }
-}
 
-// Returns true if p is inside the triangle or on a triangle's edge, false otherwise
-// Parameters in {x: 0, y:0} form
-export function _pointInTriangle (p, v1, v2, v3) {
-    // https://stackoverflow.com/questions/2049582/how-to-determine-if-a-point-is-in-a-2d-triangle
-    // contains an explanation of both this algorithm and one based on barycentric coordinates,
-    // which could be faster, but, nevertheless, it is quite similar in terms of required arithmetic operations
+    _computePolygonWidthScale (feature, viz) {
+        const SATURATION_PX = 336;
+        const diameter = Math.min(viz.strokeWidth.eval(feature), SATURATION_PX);
 
-    if (_equal(v1, v2) || _equal(v2, v3) || _equal(v3, v1)) {
-        // Avoid zero area triangle
-        return false;
+        return diameter / 2 * this.widthScale;
     }
+}
+
+const _geometryFeature = {
+    line: (feature, aabb) => {
+        const vertices = feature;
+        const numVertices = feature.length / 2;
 
-    // A point is inside a triangle or in one of the triangles edges
-    // if the point is in the three half-plane defined by the 3 edges
-    const b1 = _halfPlaneTest(p, v1, v2) < 0;
-    const b2 = _halfPlaneTest(p, v2, v3) < 0;
-    const b3 = _halfPlaneTest(p, v3, v1) < 0;
+        for (let i = 0; i < numVertices; i += 4) {
+            aabb.minx = Math.min(aabb.minx, vertices[i + 0]);
+            aabb.miny = Math.min(aabb.miny, vertices[i + 1]);
+            aabb.maxx = Math.max(aabb.maxx, vertices[i + 2]);
+            aabb.maxy = Math.max(aabb.maxy, vertices[i + 3]);
+        }
+
+        return aabb;
+    },
+
+    polygon: (feature, aabb) => {
+        const [ vertices, numVertices ] = [ feature.flat, feature.holes[0] || feature.flat.length / 2 ];
+
+        for (let k = 0; k < numVertices; k++) {
+            aabb.minx = Math.min(aabb.minx, vertices[2 * k + 0]);
+            aabb.miny = Math.min(aabb.miny, vertices[2 * k + 1]);
+            aabb.maxx = Math.max(aabb.maxx, vertices[2 * k + 0]);
+            aabb.maxy = Math.max(aabb.maxy, vertices[2 * k + 1]);
+        }
+
+        return aabb;
+    }
+};
 
-    return (b1 === b2) && (b2 === b3);
+function _getVerticesForGeometry (feature, aabb, geometryType) {
```

Comment:
This naming is not ok, this doesn't get any vertices, it returns and modifies an AABB, the initialization code of the AABB could be inside these functions to simplify things